File: vi/codes/csharp/chapter_divide_and_conquer/build_tree.cs
/**
* File: build_tree.cs
* Created Time: 2023-07-18
* Author: hpstory (hpstory1024@163.com)
*/

namespace hello_algo.chapter_divide_and_conquer;

public class build_tree {
    /* Xây dựng cây nhị phân: chia để trị */
    TreeNode? DFS(int[] preorder, Dictionary<int, int> inorderMap, int i, int l, int r) {
        // Dừng khi khoảng của cây con rỗng
        if (r - l < 0)
            return null;
        // Khởi tạo nút gốc
        TreeNode root = new(preorder[i]);
        // Truy vấn m để chia cây con trái và phải
        int m = inorderMap[preorder[i]];
        // Bài toán con: xây dựng cây con trái
        root.left = DFS(preorder, inorderMap, i + 1, l, m - 1);
        // Bài toán con: xây dựng cây con phải
        root.right = DFS(preorder, inorderMap, i + 1 + m - l, m + 1, r);
        // Trả về nút gốc
        return root;
    }

    /* Xây dựng cây nhị phân */
    TreeNode? BuildTree(int[] preorder, int[] inorder) {
        // Khởi tạo bảng băm, lưu ánh xạ từ phần tử duyệt giữa đến chỉ số
        Dictionary<int, int> inorderMap = [];
        for (int i = 0; i < inorder.Length; i++) {
            inorderMap.TryAdd(inorder[i], i);
        }
        TreeNode? root = DFS(preorder, inorderMap, 0, 0, inorder.Length - 1);
        return root;
    }

    [Test]
    public void Test() {
        int[] preorder = [3, 9, 2, 1, 7];
        int[] inorder = [9, 3, 1, 2, 7];
        Console.WriteLine("Preorder traversal = " + string.Join(", ", preorder));
        Console.WriteLine("Inorder traversal = " + string.Join(", ", inorder));

        TreeNode? root = BuildTree(preorder, inorder);
        Console.WriteLine("The constructed binary tree is:");
        PrintUtil.PrintTree(root);
    }
}

File: vi/codes/javascript/chapter_divide_and_conquer/build_tree.js
/**
 * File: build_tree.js
 * Created Time: 2023-07-30
 * Author: yuan0221 (yl1452491917@gmail.com)
 */

const { printTree } = require('../modules/PrintUtil');
const { TreeNode } = require('../modules/TreeNode');

/* Xây dựng cây nhị phân: chia để trị */
function dfs(preorder, inorderMap, i, l, r) {
    // Kết thúc khi khoảng của cây con rỗng
    if (r - l < 0) return null;
    // Khởi tạo nút gốc
    const root = new TreeNode(preorder[i]);
    // Truy vấn m để phân chia cây con trái và phải
    const m = inorderMap.get(preorder[i]);
    // Bài toán con: xây dựng cây con trái
    root.left = dfs(preorder, inorderMap, i + 1, l, m - 1);
    // Bài toán con: xây dựng cây con phải
    root.right = dfs(preorder, inorderMap, i + 1 + m - l, m + 1, r);
    // Trả về nút gốc
    return root;
}

/* Xây dựng cây nhị phân */
function buildTree(preorder, inorder) {
    // Khởi tạo bảng băm, lưu ánh xạ từ phần tử duyệt giữa sang chỉ số
    let inorderMap = new Map();
    for (let i = 0; i < inorder.length; i++) {
        inorderMap.set(inorder[i], i);
    }
    const root = dfs(preorder, inorderMap, 0, 0, inorder.length - 1);
    return root;
}

/* Đoạn mã chạy thử */
const preorder = [3, 9, 2, 1, 7];
const inorder = [9, 3, 1, 2, 7];
console.log('Preorder traversal = ' + JSON.stringify(preorder));
console.log('Inorder traversal = ' + JSON.stringify(inorder));
const root = buildTree(preorder, inorder);
console.log('The constructed binary tree is:');
printTree(root);
